applications/OptimizationApplication/custom_utilities/control/sigmoidal_projection_utils.cpp

```diff
@@ -0,0 +1,290 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   license: OptimizationApplication/license.txt
+//
+//  Main author:     Reza Najian Asl
+//                   Suneth Warnakulasuriya
+//
+
+// System includes
+#include <sstream>
+#include <type_traits>
+
+// Project includes
+#include "expression/variable_expression_io.h"
+#include "expression/variable_expression_data_io.h"
+#include "expression/container_data_io.h"
+#include "includes/define.h"
+#include "includes/variables.h"
+#include "utilities/parallel_utilities.h"
+#include "utilities/reduction_utilities.h"
+#include "utilities/variable_utils.h"
+
+// Application includes
+#include "optimization_application_variables.h"
+
+// Include base h
+#include "sigmoidal_projection_utils.h"
+
+namespace Kratos
+{
+
+///@name Kratos Classes
+///@{
+
+bool HasVectorDuplicates(
+    std::vector<double> values){
+        std::sort(values.begin(), values.end());
+        return std::adjacent_find(values.begin(), values.end()) != values.end();
+};
```

Comment:
```suggestion
bool HasVectorDuplicates(
    const std::vector<double>& rValues){
        std::set<double> unique_values(rValues.begin(), rValues.end());
        return unique_values.size() != rValues.size();
};
```

With above, this can be made const correct, and most probably faster.